### PR TITLE
unit: use more efficient sort routine

### PR DESCRIPTION
**Pull request by jbduncan**
This PR introduces a modernization change that is not necessarily related to #617 but which was spun out from my attempts at generifying the leaves of `graph`. Benchmarks to follow.

Blocked by https://github.com/gonum/gonum/pull/1940.

### CODE REVIEW

**Comment by kortschak** [comment]
This needs to pick up the changes in master.

**Comment by jbduncan** [comment]
> This needs to pick up the changes in master.

Done.

**Review by kortschak — COMMENTED** [review]
Needs evidence.

**Comment by jbduncan** [comment]
These benchmark results seem to prove that there are performance improvements:

```
goos: darwin
goarch: arm64
pkg: gonum.org/v1/gonum/unit
                             │ unit/benchmarks_dimensions_string_old.txt │ unit/benchmarks_dimensions_string_new.txt │
                             │                  sec/op                   │      sec/op        vs base                │
DimensionsString/oneDim-8                                    175.8n ± 0%         147.6n ± 0%  -16.07% (p=0.000 n=10)
DimensionsString/threeDims-8                                 488.7n ± 0%         455.1n ± 0%   -6.88% (p=0.000 n=10)
DimensionsString/allDims-8                                   1.457µ ± 0%         1.417µ ± 1%   -2.75% (p=0.000 n=10)
geomean                                                      500.3n              456.5n        -8.74%
```

**Comment by jbduncan** [comment]
Hi team, this PR is ready, so PTAL.

File: unit/unit_test.go [hunk 1]
```diff
@@ -472,3 +472,38 @@ func BenchmarkMulMapLargeSame(b *testing.B) {
...
+		MassDim:              4,
+		MoleDim:              5,
+		TemperatureDim:       6,
+		TimeDim:              7,
+		AngleDim:             8,
+	}
+
+	b.Run("oneDim", func(b *testing.B) {
+		for i := 0; i < b.N; i++ {
+			_ = oneDim.String()
```
Comment (kortschak) [inline unit/unit_test.go:496]: The assignment to `_` is a no-op here. It does not ensure that the `String` method call is not optimised away. So please remove it. If you are worried about the optimisation occurring, please add a cheap conditional (say `!= ""`) with a `b.FailNow()` in the fail case to elevate the value to an observable effect.

**Review by kortschak — COMMENTED** [review]
Please separate the benchmark addition into a separate commit, prior to the logic change so that the perf impact can be confirm eternally.

File: unit/unit_test.go (hunk 1 above)
Comment (jbduncan) [reply unit/unit_test.go]: Thanks for pointing this out. Actually I was already aware that this doesn't stop the method call from being optimised away, it was more of a habit to make it super clear that we're intentionally ignoring the result.

But it sounds sensible to use `b.FailNow()` anyway, so I've just force-pushed a change to apply this.

**Comment by jbduncan** [comment]
> Please separate the benchmark addition into a separate commit, prior to the logic change so that the perf impact can be confirm eternally.

Alrighty! Done.

I'm ready for another review whenever you are.

**Review by kortschak — APPROVED** [review]
Thanks

**Comment by jbduncan** [comment]
@kortschak yw! And thank you for being so patient with me, with all my PRs recently. :)

**Pull request merged**